File: Codesignal/Eric/73.SwitchLights.go
package main

import (
	"fmt"
)

func solution(a []int) []int {
	result := make([]int, 0)
	for i, v := range a {
		acc := 0
		for j := i; j < len(a); j++ {
			acc += a[j]
		}
		val := v
		if acc%2 == 1 {
			val = v - 1
			if val < 0 {
				val = 1
			}
		}

		result = append(result, val)
	}
	return result
}

func main() {
	fmt.Println(solution([]int{1, 1, 1, 1, 1})) // [0,1,0,1,0]
}

File: Codesignal/Eric/73.SwitchLights.py
def solution(a):
    result = []
    for i in range(len(a)):
        acc = 0
        for j in range(i, len(a)):
            acc += a[j]
        result.append(a[i] if acc % 2 == 0 else abs(a[i]-1))
    return result
            
def solution2(a):
    return [(x + sum(a[i:]))%2 for i,x in enumerate(a)]

def solution3(a):
    flips = 0
    for i in range(len(a)-1,-1,-1):
        if a[i]==1:
            flips +=1
        a[i] = (a[i]+flips)%2
    return a

def solution4(a):
    for i in range(len(a)):
        if a[i]:
            for j in range(i+1):
                a[j]=1-a[j]
    return a

print(solution2([1, 1, 1, 1, 1])) # [0, 1, 0, 1, 0]